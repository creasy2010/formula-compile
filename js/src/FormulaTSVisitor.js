// Generated from /Users/dong/yzfworkbench/formula-compile/java/FormulaTS.g4 by ANTLR 4.8
// jshint ignore: start
var antlr4 = require('antlr4/index');

// This class defines a complete generic visitor for a parse tree produced by FormulaTSParser.

function FormulaTSVisitor() {
	antlr4.tree.ParseTreeVisitor.call(this);
	return this;
}

FormulaTSVisitor.prototype = Object.create(antlr4.tree.ParseTreeVisitor.prototype);
FormulaTSVisitor.prototype.constructor = FormulaTSVisitor;

// Visit a parse tree produced by FormulaTSParser#formulaUnit.
FormulaTSVisitor.prototype.visitFormulaUnit = function(ctx) {
  return this.visitChildren(ctx);
};


// Visit a parse tree produced by FormulaTSParser#formulaFunction.
FormulaTSVisitor.prototype.visitFormulaFunction = function(ctx) {
  return this.visitChildren(ctx);
};


// Visit a parse tree produced by FormulaTSParser#formulaParams.
FormulaTSVisitor.prototype.visitFormulaParams = function(ctx) {
  return this.visitChildren(ctx);
};


// Visit a parse tree produced by FormulaTSParser#formulaParam.
FormulaTSVisitor.prototype.visitFormulaParam = function(ctx) {
  return this.visitChildren(ctx);
};


// Visit a parse tree produced by FormulaTSParser#formulaExpress.
FormulaTSVisitor.prototype.visitFormulaExpress = function(ctx) {
  return this.visitChildren(ctx);
};


// Visit a parse tree produced by FormulaTSParser#formulaBracketExpress.
FormulaTSVisitor.prototype.visitFormulaBracketExpress = function(ctx) {
  return this.visitChildren(ctx);
};


// Visit a parse tree produced by FormulaTSParser#formulaIfFunction.
FormulaTSVisitor.prototype.visitFormulaIfFunction = function(ctx) {
  return this.visitChildren(ctx);
};


// Visit a parse tree produced by FormulaTSParser#formulaRefTemplateFunction.
FormulaTSVisitor.prototype.visitFormulaRefTemplateFunction = function(ctx) {
  return this.visitChildren(ctx);
};


// Visit a parse tree produced by FormulaTSParser#formulaAssignExpress.
FormulaTSVisitor.prototype.visitFormulaAssignExpress = function(ctx) {
  return this.visitChildren(ctx);
};


// Visit a parse tree produced by FormulaTSParser#formulaFunctionName.
FormulaTSVisitor.prototype.visitFormulaFunctionName = function(ctx) {
  return this.visitChildren(ctx);
};


// Visit a parse tree produced by FormulaTSParser#formulaParamConst.
FormulaTSVisitor.prototype.visitFormulaParamConst = function(ctx) {
  return this.visitChildren(ctx);
};


// Visit a parse tree produced by FormulaTSParser#formulaParamNum.
FormulaTSVisitor.prototype.visitFormulaParamNum = function(ctx) {
  return this.visitChildren(ctx);
};


// Visit a parse tree produced by FormulaTSParser#formulaParamString.
FormulaTSVisitor.prototype.visitFormulaParamString = function(ctx) {
  return this.visitChildren(ctx);
};


// Visit a parse tree produced by FormulaTSParser#formulaParamNull.
FormulaTSVisitor.prototype.visitFormulaParamNull = function(ctx) {
  return this.visitChildren(ctx);
};


// Visit a parse tree produced by FormulaTSParser#formulaCELLLoc.
FormulaTSVisitor.prototype.visitFormulaCELLLoc = function(ctx) {
  return this.visitChildren(ctx);
};


// Visit a parse tree produced by FormulaTSParser#formulaCELLTPLLoc.
FormulaTSVisitor.prototype.visitFormulaCELLTPLLoc = function(ctx) {
  return this.visitChildren(ctx);
};


// Visit a parse tree produced by FormulaTSParser#skipFuncLBracket.
FormulaTSVisitor.prototype.visitSkipFuncLBracket = function(ctx) {
  return this.visitChildren(ctx);
};


// Visit a parse tree produced by FormulaTSParser#skipFuncRBracket.
FormulaTSVisitor.prototype.visitSkipFuncRBracket = function(ctx) {
  return this.visitChildren(ctx);
};


// Visit a parse tree produced by FormulaTSParser#skipParamComma.
FormulaTSVisitor.prototype.visitSkipParamComma = function(ctx) {
  return this.visitChildren(ctx);
};



exports.FormulaTSVisitor = FormulaTSVisitor;