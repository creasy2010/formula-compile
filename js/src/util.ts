import {CommonToken, ParserRuleContext, Token} from "antlr4";
import {TerminalNodeImpl} from "antlr4/tree/Tree";

/**
 * @desc
 *
 * @使用场景
 *
 * @coder.yang2010@gmail.com
 * @Date    2020/7/9
 **/



export function toJSON(tree:ParserRuleContext){
  let result  ={};
  traverse(tree,result);
  return result;
}


export function getTokenInfo(token:Token){
  return {
    // type:token.type,
    line:token.line,
    start:token.start,
    stop:token.stop,
    column:token.column,
  };
}


export function getRangeInfo(ctx:ParserRuleContext){

  // return "todo";
  return {
    depth:ctx.depth(),
    start: getTokenInfo(ctx.start),
    stop: getTokenInfo(ctx.stop),
    _source: ctx.getText(),
  }
}

export function getTerminalNodeInfo(terminalNode:TerminalNodeImpl){
  let token  = terminalNode.getSymbol();

  return {
    "!":"TerminalNodeImpl",
    value:terminalNode.getText(),
    range:{
      // type:token.type,
      line:token.line,
      startIndex:token.start,
      stopIndex:token.stop,
      column:token.column,
      _source:terminalNode.getText()
    }
  }
}
function traverse(tree:ParserRuleContext,result){

  if(tree instanceof TerminalNodeImpl){
    let token = tree.getSymbol();
    // result.token=token;
    result.range={
      type:token.type,
      line:token.line,
      startIndex:token.start,
      stopIndex:token.stop,
      column:token.column,
    };
    result.text =token.text;
  }else{
    let children = [];
    let name  = Object.getPrototypeOf(tree).constructor.name.replace(/Context$/,"");
    result[name] = children;
    //@ts-ignore
    if(tree?.children?.length>0){
      //@ts-ignore
      for (let i = 0, iLen = tree.children?.length; i < iLen; i++) {
        //@ts-ignore
        let child = tree.children[i];
        let childInfo={
          type:name,
        };
        children.push(childInfo);
        traverse(child,childInfo);
      }
    }
  }

}

export const formulaFormat = (()=>{
  const format = (ast)=>{
    let rawFormula = '';
    if (ast && Object.prototype.toString.call(ast.formulas) === "[object Array]") {
      ast.formulas.forEach((formula, index)=>{
        rawFormula += formatAtomicFormula(formula)
        if (index !== ast.formulas.length-1) {
          rawFormula += ';'
        }
      })
    }
    return rawFormula;
  }
  const formatAtomicFormula = (obj, newLine=false, space='\t')=>{
    let rawAtomicFormula = '';
    if (newLine) {
      // rawAtomicFormula += '\n' + space;
      // space += '\t';  //处理缩进层级显示
    }
    if(!obj){
      return rawAtomicFormula;
    }
    if (typeof obj === 'string') {
      return rawAtomicFormula;
    }
    const name = obj.name;
    const type = obj['!'];
    return rawAtomicFormula + formulaStrategy[type](obj, newLine, space);
  }

  var formulaStrategy = {
    FormulaRefTemplateFunction: (obj, newLine, space)=>{
      return `RefTemplate(${obj.refSheet.value},${obj.refCell.value})`;
    },
    FormulaIfFunction: (obj, newLine, space)=>{
      let rawAtomicFormula = '';
      rawAtomicFormula += `IF(`;
      const leftExp = formatAtomicFormula(obj.condition.content, false, space);
      const thenExp = formatAtomicFormula(obj.thenStatement.content, true, space);
      const elseExp = formatAtomicFormula(obj.elseStatement.content, true, space);
      rawAtomicFormula += `${leftExp},${thenExp},${elseExp})`;
      return rawAtomicFormula;
    },
    FormulaBracketExpress: (obj, newLine, space)=>{
      let rawAtomicFormula = '';
      rawAtomicFormula += `(`;
      obj.express.forEach(exp=>{
        rawAtomicFormula += formatAtomicFormula(exp, false, space);
      })
      rawAtomicFormula += `)`;
      return rawAtomicFormula;
    },
    FormulaBinaryOperationExpress: (obj, newLine, space)=>{
      let rawAtomicFormula = '';
      rawAtomicFormula += `${formatAtomicFormula(obj.leftExp, false, space)}`;
      if(obj.operate){
        if (obj.operate.value === '+' || obj.operate.value === '-' || obj.operate.value === '*' || obj.operate.value === '/') {
          // rawAtomicFormula += '\n\t';
        }
        rawAtomicFormula += `${obj.operate.value}`;
      }
      if(obj.rightExp){
        rawAtomicFormula += `${formatAtomicFormula(obj.rightExp, false, space)}`;
      }
      return rawAtomicFormula;
    },
    FormulaParamString: (obj, newLine, space)=>{
      return `'${obj.value}'`;
    },
    FormulaParamConst: (obj, newLine, space)=>{
      return `${obj.value}`;
    },
    FormulaParamNum: (obj, newLine, space)=>{
      return `${obj.value}`;
    },
    FormulaExpress: (obj, newLine, space)=>{
      let rawAtomicFormula = '';
      obj.express.forEach(exp=>{
        rawAtomicFormula += formatAtomicFormula(exp, newLine, space);
      })
      return rawAtomicFormula;
    },
    FormulaCELLLoc: (obj, newLine, space)=>{
      let rawAtomicFormula = '';
      if (obj.sheetName) {
        rawAtomicFormula += `{${obj.sheetName}}!`;
      }
      if (obj.cellLoc) {
        rawAtomicFormula += `[${obj.cellLoc}]`;
      }
      return rawAtomicFormula;
    },
    FormulaFunction: (obj, newLine,space)=>{
      let rawAtomicFormula = '';
      const name = obj.name;      
      if (typeof name === 'string') {
        rawAtomicFormula += `${name}(`;
        obj.params.forEach((param, paramIndex)=>{
            const paramRes = formatAtomicFormula(param.content, false, space);
            if (obj.params.length === 1) {
              rawAtomicFormula += `${paramRes}`;
            }else if (paramIndex===0) {
              rawAtomicFormula += `${paramRes},`;
            }else if(paramIndex === obj.params.length -1){
              rawAtomicFormula += `${paramRes}`;
            }else{
              rawAtomicFormula += `${paramRes},`;
            }
        })
        rawAtomicFormula += `)`;
      }
      return rawAtomicFormula;
    }
  }

  return format;
})()