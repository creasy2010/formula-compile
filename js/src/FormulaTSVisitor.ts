/**
 * @desc
 *
 * @使用场景
 *
 * @coder.yang2010@gmail.com
 * @Date    2020/7/9
 **/
import {ParseTreeVisitor, TerminalNodeImpl} from 'antlr4/tree/Tree';
import {ParserRuleContext} from 'antlr4';
import { getRangeInfo, getTerminalNodeInfo, getTokenInfo} from './util';

export interface FormulaAst {
  [name: string]: any;
}

/**
 * 转换为ast
 */
export class FormulaTSVisitor extends ParseTreeVisitor {

  constructor() {
    super();
  }

 async toAst(formula: string): Promise<FormulaAst> {
    let parse=require('./index').parseFormula;
    // console.time('vistor:parse');
    let context = await parse(formula);
   // console.timeEnd('vistor:parse');
     return this.visit(context);
  }

  visitFormulaUnit(ctx: ParserRuleContext) {

    return {
      '!': 'FormulaUnit',
      range: getRangeInfo(ctx),
      formulas: this.visitChildren(ctx).filter(item=>{
        return !!item;
      }),
    };
  }

  // Visit a parse tree produced by FormulaTSParser#formulaIfFunction.
  visitFormulaIfFunction (ctx) {
    let [condition,thenStatement,elseStatement]  = this.visit(ctx.children[2].children).filter(item=>!!item);

    return {
      '!': 'FormulaIfFunction',
      range: getRangeInfo(ctx),
      name:ctx.children[0].getText(),
      condition,
      thenStatement,
      elseStatement
    };
  };

  // Visit a parse tree produced by FormulaTSParser#formulaRefTemplateFunction.
  visitFormulaRefTemplateFunction(ctx) {
    return {
      '!': 'FormulaRefTemplateFunction',
      range: getRangeInfo(ctx),
      name:ctx.children[0].getText(),
      refSheet:{
        range:getTerminalNodeInfo(ctx.children[2]),
        value:ctx.children[2].getText()
      },
      refCell:{
        range:getTerminalNodeInfo(ctx.children[4]),
        value:ctx.children[4].getText()
      }
    };
  };

  visitFormulaAssignExpress(ctx){
    return {
      '!': 'FormulaAssignExpress',
      range: getRangeInfo(ctx),
      varName:ctx.children[0].getText(),
      express:this.visit(ctx.children[2])
    };
  }

  // Visit a parse tree produced by FormulaTSParser#formulaFunction.
  visitFormulaFunction(ctx) {
    let params=[];
    if( ctx.children.length > 3 && ctx.children[2].children) {
      params = ctx.children[2].children?.map(item=>{
        if(item instanceof TerminalNodeImpl) {
          return getTerminalNodeInfo(item);
        } else {
          return this.visit(item)
        }
      }).filter(item=>!!item)
    }

    return {
      '!': 'FormulaFunction',
      range: getRangeInfo(ctx),
      name:this.visit(ctx.children[0]),
      params
    };
  }

  // Visit a parse tree produced by FormulaTSParser#formulaParams.
  visitFormulaParams(ctx) {
    return this.visitChildren(ctx.children)
  }

  // Visit a parse tree produced by FormulaTSParser#formulaParam.
  visitFormulaParam(ctx) {
    return {
      '!': 'FormulaParam',
      range: getRangeInfo(ctx),
      content:this.visit(ctx.children[0])
    }
  }

  // Visit a parse tree produced by FormulaTSParser#formulaFunctionName.
  visitFormulaFunctionName(ctx:ParserRuleContext) {
    return ctx.getText();
  }

  // Visit a parse tree produced by FormulaTSParser#formulaParamConst.
  visitFormulaParamConst(ctx:ParserRuleContext) {
    return {
      '!': 'FormulaParamConst',
      range: getRangeInfo(ctx),
      value:ctx.getText()
    }
  }

  // Visit a parse tree produced by FormulaTSParser#formulaParamNum.
  visitFormulaParamNum(ctx:ParserRuleContext) {

    return {
      '!': 'FormulaParamNum',
      range: getRangeInfo(ctx),
      value:ctx.getText()
    }
  }

  // Visit a parse tree produced by FormulaTSParser#formulaParamString.
  visitFormulaParamString(ctx:ParserRuleContext) {
    let content =ctx.getText();
    return {
      '!': 'FormulaParamString',
      range: getRangeInfo(ctx),
      value:content.substring(1,content.length-1)
    }
  }


// Visit a parse tree produced by FormulaTSParser#formulaParamNull.
  visitFormulaParamNull(ctx) {
    return {
      '!': 'FormulaParamNull',
      range: getRangeInfo(ctx),
      value:ctx.getText()
    };
  }

  // Visit a parse tree produced by FormulaTSParser#formulaCELLLoc.
  visitFormulaCELLLoc(ctx:ParserRuleContext) {
    let cellLoc = null , sheetName=null;
     if(ctx.getChildCount()===1){
       let cellStr:string = ctx.getChild(0).getText();
       cellLoc=cellStr.substr(1,cellStr.length-2);
     } else if(ctx.getChildCount()===2){
        let sheetStr:string = ctx.getChild(0).getText();
        sheetName=sheetStr.substr(1,sheetStr.length-3);
       let cellStr:string = ctx.getChild(1).getText();
       cellLoc=cellStr.substr(1,cellStr.length-2);
     }else{
       throw new Error('未知语法格式');
     }

    return {
      '!': 'FormulaCELLLoc',
      range: getRangeInfo(ctx),
      cellLoc,
      sheetName,
    };


  }
//公式模板引用位置.
  visitFormulaCELLTPLLoc(ctx) {

    let cellLoc=ctx.children[0].getText().replace("_@","");
    return {
      '!': 'FormulaCELLTPLLoc',
      range: getRangeInfo(ctx),
      cellLoc,
    };
  };

  // Visit a parse tree produced by FormulaTSParser#formulaExpress.
  visitFormulaExpress(ctx) {
    // TODO dong 2020/7/9 ast 解析的有问题.

    if(ctx.getChildCount()===1) {
      return {
        '!': 'FormulaExpress',
        range: getRangeInfo(ctx),
        express:this.visit(ctx.children),
      };
``    } else if(ctx.getChildCount()===5 && ctx.children[1].getText()==='?' && ctx.children[3].getText()===':') {
      //三元表达式;

      return  {
          '!': 'FormulaTernaryExpression',
          range: getRangeInfo(ctx),
          condition:this.visit(ctx.children[0]),
          trueExp:this.visit(ctx.children[2]),
          falseExp:this.visit(ctx.children[4]),
        }
    } else if(ctx.getChildCount()===3) {
      //二元运算;
      // if(!this.visit(ctx.children[0])) {
      //   debugger;
      // }
      return {
        '!': 'FormulaBinaryOperationExpress',
        range: getRangeInfo(ctx),
        leftExp:this.visit(ctx.children[0]),
        operate:getTerminalNodeInfo(ctx.children[1]),
        rightExp:this.visit(ctx.children[2]),
      };
    } else {
      throw new Error('未知表达式'+ ctx.getText());
    }

  }


// Visit a parse tree produced by FormulaTSParser#formulaBracketExpress.
  visitFormulaBracketExpress (ctx) {
    let express = ctx.children.slice(1,ctx.children.length-1).map(item=>this.visit(item));

    return {
      '!': 'FormulaBracketExpress',
      range: getRangeInfo(ctx),
      express:express,
    };
  };



  // Visit a parse tree produced by FormulaTSParser#formulaOperation.
  visitFormulaOperation(ctx) {
    return {
      '!': 'FormulaOperation',
      range: getRangeInfo(ctx),
      value:ctx.getText()
    };
  }


// Visit a parse tree produced by FormulaTSParser#skipFuncLBracket.
  visitSkipFuncLBracket(ctx) {
    return ;
  };


// Visit a parse tree produced by FormulaTSParser#skipFuncRBracket.
  visitSkipFuncRBracket(ctx) {
    return ;
  };


// Visit a parse tree produced by FormulaTSParser#skipParamComma.
  visitSkipParamComma(ctx) {
    return ;
  };


}
