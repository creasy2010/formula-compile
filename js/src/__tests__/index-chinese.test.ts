/**
 * @desc
 *
 * @使用场景
 *
 * @coder.yang2010@gmail.com
 * @Date    2020/7/9
 **/
import {FormulaTSVisitor} from "../FormulaTSVisitor";
import {toJSON, formulaFormat, toggleRangeInfo} from "../util";

toggleRangeInfo();

describe('中英文场景', async() => {

  it('visitor处理',async()=>{
    let visitor  = new FormulaTSVisitor();
    for (let i = 0, iLen = formulas.length; i < iLen; i++) {
      let formula = formulas[i];
      if(!formula) {
        continue;
      }
      let timeFLag =`公式[${i}]耗时 长度:${formula?.length}`;
      console.time(timeFLag);
      try {

        let ast  =await visitor.toAst(formula);
        expect(ast).toMatchSnapshot('测试公式formulas['+i+']');
      } catch (err) {
        expect(jest.fn()).not.toBeCalled();
      }
      console.timeEnd(timeFLag);
    }
  })

});


let formulas:string[] = [
  `if（1+1>2，0，1）`
];
