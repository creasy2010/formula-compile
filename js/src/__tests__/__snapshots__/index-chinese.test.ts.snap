// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`中英文场景 visitor处理: 测试公式formulas[0] 1`] = `
Object {
  "!": "FormulaUnit",
  "formulas": Array [
    Object {
      "!": "FormulaExpress",
      "express": Array [
        Object {
          "!": "FormulaFunction",
          "name": "if",
          "params": Array [
            Object {
              "!": "FormulaParam",
              "content": Object {
                "!": "FormulaParamNum",
                "range": Object {
                  "_source": "1",
                },
                "value": "1",
              },
              "range": Object {
                "_source": "1",
              },
            },
            Object {
              "!": "FormulaParam",
              "content": Object {
                "!": "FormulaBinaryOperationExpress",
                "leftExp": Object {
                  "!": "FormulaExpress",
                  "express": Array [
                    Object {
                      "!": "FormulaParamNum",
                      "range": Object {
                        "_source": "+1",
                      },
                      "value": "+1",
                    },
                  ],
                  "range": Object {
                    "_source": "+1",
                  },
                },
                "operate": Object {
                  "!": "TerminalNodeImpl",
                  "range": Object {
                    "_source": ">",
                  },
                  "value": ">",
                },
                "range": Object {
                  "_source": "+1>2",
                },
                "rightExp": Object {
                  "!": "FormulaExpress",
                  "express": Array [
                    Object {
                      "!": "FormulaParamNum",
                      "range": Object {
                        "_source": "2",
                      },
                      "value": "2",
                    },
                  ],
                  "range": Object {
                    "_source": "2",
                  },
                },
              },
              "range": Object {
                "_source": "+1>2",
              },
            },
            Object {
              "!": "FormulaParam",
              "content": Object {
                "!": "FormulaParamNum",
                "range": Object {
                  "_source": "0",
                },
                "value": "0",
              },
              "range": Object {
                "_source": "0",
              },
            },
            Object {
              "!": "FormulaParam",
              "content": Object {
                "!": "FormulaParamNum",
                "range": Object {
                  "_source": "1",
                },
                "value": "1",
              },
              "range": Object {
                "_source": "1",
              },
            },
          ],
          "range": Object {
            "_source": "if（1+1>2，0，1）",
          },
        },
      ],
      "range": Object {
        "_source": "if（1+1>2，0，1）",
      },
    },
  ],
  "range": Object {
    "_source": "if（1+1>2，0，1）",
  },
}
`;
